[WIP] Initial GIBCT react app

typesend opened the pull request:
Sharing my #3960 progress so far for review!

This is my first react application so I have definitely experienced a fair amount of "two-steps-forward, one-step back", but I feel like this is finally evolving in the right direction.

![image](https://cloud.githubusercontent.com/assets/8130/20810844/5f5fcc5a-b7d9-11e6-94ad-c93d37b23a9b.png)

FYI, this includes code from #4006

@emilyville @ayaleloehr @plusjeff @mphprogrammer @patrickvinograd

ben-USDS-damman committed:
Initial GIBCT react app

ben-USDS-damman committed:
Introduce refactored Estimator

ben-USDS-damman committed:
exclude GIBCT from production builds

austin-martinez commented on src/js/gi-bill-comparison-tool/components/AboutYourSchoolFields.jsx:
```diff
@@ -0,0 +1,66 @@
+import React from 'react';
+import Dropdown from './Dropdown';
+import Modal from '../../common/components/Modal';
```
Don't see this or Dropdown used here

ben-USDS-damman committed:
re-enable _gi-modal scss

ben-USDS-damman committed:
removed unused imports

austin-martinez commented on src/js/gi-bill-comparison-tool/components/Dropdown.jsx:
```diff
@@ -0,0 +1,79 @@
+import React from 'react';
+
+class Dropdown extends React.Component {
```
I'm not too familiar with all of our common React components, but we do have one for selects, could that be used here, or extended to add in any missing functionality?

typesend replied:
There is `<ErrorableSelect/>` but I didn't want to bite off trying to reuse that before I knew what that css style implications of that would be. Certainly open to reusing/extending that if people agree it makes sense (as a future enhancement.)

jbalboni replied:
All of our components should use the USWDS styles, so it doesn't look like it would change much, style-wise. But waiting to incorporate them until the rewrite is more settled seems reasonable to me.

austin-martinez commented on src/js/gi-bill-comparison-tool/containers/LandingPage.jsx:
```diff
@@ -0,0 +1,54 @@
+          <li><a href="/">Home</a></li>
+          <li><a href="/education/">Education Benefits</a></li>
+          <li><a href="/education/gi-bill/">GI Bill</a></li>
+          <li className="active">GI Bill® Comparison Tool</li>
```
Why the ® here after GI Bill but not in the preceding breadcrumb? Had no idea that was a registered trademark, do we need to denote that? http://benefits.va.gov/GIBILL/Trademark_Terms_of_Use.asp this looks like it would only be necessary for third parties referring to GI Bill, but not the VA itself

typesend replied:
I agree that it is unnecessary. Removing.

ben-USDS-damman committed:
fixes to make eslink happier

ben-USDS-damman committed:
remove trademark registration symbol

ben-USDS-damman committed:
AboutYourSchoolFields === and duplicate prop corrections

ben-USDS-damman committed:
eslint corrections

jbalboni commented on src/js/gi-bill-comparison-tool/components/Dropdown.jsx:
```diff
@@ -0,0 +1,79 @@
+        <select id={this.props.identifier}
+            name={this.props.identifier.replace(/\-/g, '_')}
+            alt={this.props.alt}
+            onChange={(e) => { this.props.onChange(e) && this.setState({ value: event.target.value }); }}
```
We've been writing our form input components as [controlled components](https://facebook.github.io/react/docs/forms.html#controlled-components), where we pass the value down from a parent component (and changes back up with `onChange` hooks like you have here). You can check out how `ErrorableSelect` works to see it. It'll make it easier to migrate to our common components later if you use the same style here (or give `ErrorableSelect` a shot now).

jbalboni commented on src/js/gi-bill-comparison-tool/components/Dropdown.jsx:
```diff
@@ -0,0 +1,79 @@
+    return this.props.children;
+  }
+
+  renderDropdownOptions(optionsArray) {
```
Is there a reason for passing the options array as an array of strings, instead of an array of objects? It seems like it's making this more complicated.

If you passed in options arrays like this:

```
[
  { value: 'yes', label: 'Eligilble for Post-9/11 Gi Bill' },
  { value: 'no', label: 'Not Eligilble for Post-9/11 Gi Bill'}
]
```

This function can be one line:

`return optionsArray.map(({ value, label }) => <option key={value} value={value}>{label}</option>);`

typesend replied:
Although it does increase the complexity of this one clear-purpose function, it really simplifies the code where these Dropdowns are used.

ben-USDS-damman committed:
Renaming gi-comparison-tool directories and namespacing to simply gi

ben-USDS-damman committed:
adding myself to humans.txt

ben-USDS-damman committed:
remove additonal resources sidebar, copy over calculator

ben-USDS-damman committed:
moving inline styles

ben-USDS-damman committed:
profile components

ben-USDS-damman committed:
page titles

ben-USDS-damman committed:
function reordering, modals render only if labels shown

ben-USDS-damman committed:
breadcrumbs component

ben-USDS-damman committed:
add school mock data

ben-USDS-damman committed:
moar profile components

ben-USDS-damman committed:
css span class unique to each Page component

ben-USDS-damman committed:
profile vet summary area

ben-USDS-damman committed:
css experimentation

ben-USDS-damman committed:
profile estimator

ben-USDS-damman committed:
first passing estimator tests

ben-USDS-damman committed:
profile estimator complete and passing tests

ben-USDS-damman committed:
profile page modals and other display improvements

ben-USDS-damman committed:
consistent span containers for modals

ben-USDS-damman committed:
adding d3 and graph js

ben-USDS-damman committed:
RetentionRates and first graph placeholder

ben-USDS-damman committed:
outcomes components with graph placeholders

ben-USDS-damman committed:
Profile History and its modals

ben-USDS-damman committed:
ProfileHistory number formatting

ben-USDS-damman committed:
Outcomes and Complaints profile component improvements

ben-USDS-damman committed:
Caculator Modals

ben-USDS-damman committed:
Calculator component inputs and outputs, old onclick comments

ben-USDS-damman committed:
refactoring calculator class

ben-USDS-damman committed:
calculator reliant upon component state

ben-USDS-damman committed:
calculations

ben-USDS-damman committed:
partial eslint-driven syntax changes

ben-USDS-damman committed:
all gi components eslint syntax fixed

ben-USDS-damman committed:
container eslint syntax fixes

ben-USDS-damman committed:
automated eslint fixes for calc and estimator

ben-USDS-damman committed:
eslint calc fixing

ben-USDS-damman committed:
moar eslint fixes

ben-USDS-damman committed:
npm-shrinkwrap

ben-USDS-damman committed:
camelCasing all the things

ben-USDS-damman committed:
all eslint fixes complete

ben-USDS-damman committed:
okay moar syntax fixes made after all

ben-USDS-damman committed:
calculator spec and file tidying

- remove deprecated schema submodule
- removed keep file
- calculator is not a component
- some eslint fixes in tests

ben-USDS-damman committed:
rename calculator.js back to .jsx

- apparently calculator will only load as a module if jsx extension used

ben-USDS-damman committed:
fix <select> name references of dependendent dropdowns

ben-USDS-damman committed:
fix estimator and calculator module name case

ben-USDS-damman committed:
calc and estmator test name fix

U-DON commented on src/js/gi/components/AboutYourselfFields.jsx:
```diff
@@ -0,0 +1,358 @@
+            onChange={this.handleDropdownChange}
+            showLabel={this.props.labels}/>
+
+        <Modal onClose={() => {this.toggleModalDisplay('militaryStatus');}} visible={!!this.state.modals.militaryStatus}>
```
Not sure if this is considered more idiomatic or not, but what I like to do is pass something like:

`this.toggleModalDisplay.bind(this, 'militaryStatus')`

Should end up being the same as creating the anonymous function.

typesend replied:
Nice! This is way better than the anonymous function garbage.

U-DON commented on src/js/gi/components/AdditionalResources.jsx:
```diff
@@ -0,0 +1,59 @@
+
+  render() {
+    return (
+      <span>
```
This should probably be a `div`.

typesend replied:
I was using `<span>` just to denote this parent tag was needed by JSX but not in the original project code. Agree, though, and changing this across all components where it was done.

U-DON commented on src/js/gi/components/Breadcrumbs.jsx:
```diff
@@ -0,0 +1,45 @@
+import React from 'react';
+
+class Breadcrumbs extends React.Component {
```
We have a Breadcrumbs component in `rx` as well. Probably not for this PR, but it might be useful to pull both out into `common` at some point.

typesend replied:
Noted.

U-DON commented on src/js/gi/components/Dropdown.jsx:
```diff
@@ -0,0 +1,75 @@
+  // renders a list of options defined as an array
+  // in the form of [value1, label1, value2, label2...]
+  renderDropdownOptions(optionsArray) {
+    const options = Object.assign([], optionsArray);
```
I don't think `Object.assign` is intended for cloning arrays. Instead, maybe:

`const options = optionsArray.slice();`

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice

typesend replied:
Good catch

U-DON commented on src/js/gi/components/SearchResult.jsx:
```diff
@@ -0,0 +1,153 @@
+    this.estimate();
+
+    return (
+      <div id={this.props.facility_code} className="school_summary large-4 medium-5-center small-12-center columns">
```
The props are all in camelCase, so probably meant `this.props.facilityCode`. Same with `cautionFlag` below.

typesend replied:
Yep, fixing

ben-USDS-damman committed:
npm-shrinkwrap: rm version for vets schema

U-DON commented on src/js/gi/components/SearchResult.jsx:
```diff
@@ -0,0 +1,153 @@
+  // WIP
+  estimate() {
+    // get current values about the user
+    const e = this.props.estimator;
```
I see that it's a WIP, but in case you didn't already know, it's not considered good practice to modify props.

U-DON commented on src/js/gi/components/profile/GraduationRates.jsx:
```diff
@@ -0,0 +1,60 @@
+    const isNumeric = (n) => { return !Number.isNaN(parseFloat(n)); };
+    const isaV = isNumeric(school.graduationRateVeteran);
+    const isaC = isNumeric(school.graduationRateAllStudents);
+    const heading = <h3>Graduation Rate<a onClick={() => {this.props.toggleModalDisplay('gradrates');}} className="info-icons"><i className="fa fa-info-circle info-icons outcomes-learnmore"></i></a></h3>;
```
Would be nice to break this up into multiple lines for readability.

U-DON commented on src/js/gi/containers/LandingPageForm.jsx:
```diff
@@ -0,0 +1,55 @@
+  render() {
+    return (
+      <div className="row">
+        <form action="/gi-bill-comparison-tool/institutions/search">
```
Does this end up doing a full page refresh? If it's hooked up in the routes, you should have the `onSubmit` handler do something like a `history.push` instead, so it can render in place.

U-DON commented on src/js/gi/components/Breadcrumbs.jsx:
```diff
@@ -0,0 +1,45 @@
+  renderNamed(props) {
+    return (
+      <span>
+        <li><a href="/gi-bill-comparison-tool/">GI Bill Comparison Tool</a></li>
```
This should probably be a `Link` from `react-router`. Any of this app's routes should use `Link` rather than `a`, or else the links will load the full page.

Also, to avoid prefixing all the internal links with `/gi-bill-comparison-tool`, you can do something like this: https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/js/rx/rx-entry.jsx#L16

ben-USDS-damman committed:
changed component root tags from <span> to <div> where applicable

typesend commented on src/js/gi/components/AboutYourselfFields.jsx:
```diff
@@ -0,0 +1,358 @@
+  }
+
+  renderVREBenefitsMessage() {
+    // onClick="track('Tool Tips', 'Voc Rehab Warning');"
```
This is a reminder for how click tracking was implemented in the rails app

ben-USDS-damman committed:
change some click handlers to use bind

ben-USDS-damman committed:
use Array.slice() instead of Object.assign

ben-USDS-damman committed:
fix prop names

ben-USDS-damman committed:
grad rates: heading multi-line code, click handler simplification

ben-USDS-damman committed:
label styles and onlick simplifying

ben-USDS-damman committed:
switch component props.estimator to state.estimator

typesend commented on src/js/gi/components/AboutYourSchoolFields.jsx:
```diff
@@ -0,0 +1,61 @@
+              <input type="radio" id="online-classes-all"
+                  name="online_classes" value="yes"
+                  defaultChecked={this.props.onlineClasses && this.props.onlineClasses === 'all'}/>
+              <label className="radio" htmlFor="online-classes-all" style={{ display: 'inline-block' }}>All</label>
```
Turns out `inline-block` is needed so I moved it into a stylesheet. (These are old styles; I've spent as little time as possible cleaning up old css.)

ben-USDS-damman committed:
eslint fix

U-DON commented on script/build.js:
```diff
@@ -116,6 +116,13 @@ smith.destination(`../build/${options.buildtype}`);
+const ignore = require('metalsmith-ignore');
+const ignoreList = [];
+if (options.buildtype === 'production') {
+  ignoreList.push('gi/*');
```
I think this uses the relative path from the `content` folder, so it should probably be `'gi-bill-comparison-tool/*`.

typesend commented:
Thank you reviewers!
